extendr-api/src/robj/mod.rs

```diff
@@ -734,65 +734,125 @@ where
 {
     fn as_typed_slice(&self) -> Option<&'a [T]>
     where
-        Self: 'a,
-    {
-        None
-    }
+        Self: 'a;
 
     fn as_typed_slice_mut(&mut self) -> Option<&'a mut [T]>
     where
-        Self: 'a,
-    {
-        None
+        Self: 'a;
+}
+
+trait SExpAccessor {
```

Comment:
Needs docs. Also try to think of a more precise name for this. `SexpAccessibleAs?` Also, can this be generic rather than using an associated type?

Reply:
Unfortunately, it has to be an associated type.

Reply:
I don't know what to describe it as..

Reply:
To begin with, I just wrote the trait necessary to distinguish between the types and in order to make the non-macro impl work. I didn't intend on diving into what this is and what it does.
But I've gained some understanding: Basically it unsafely-casts a pointer from an opaque pointer, to a rust-typed pointer.
I've renamed things to reflect that. 
A vector or not will be hidden behind a pointer regardless; In fact everything is hidden behind this type anyways, so there isn't really a need to write that. I've put in a reference to `Rtype` which shows that there are `XP` (C-ints that are ids for r-types) for everything. Incidentally, I've renamed `XP` to `R_TYPE_ID`.
Hope that's good enough for this PR. I don't mind exploring other things now that they come up, but I'm not sure this should always happen in PRs. This time I think it was fine.